doc/developer/design/20230519_statement_logging.md

```diff
@@ -0,0 +1,165 @@
+  with other relevvant information (e.g., timestamp information, time
+  to return, current cluster, current user, plan information, etc.)
+* (Stretch goal) unprivileged users should be able to access a view
+  that filters `mz_statement_history` to only surface queries that
+  that user has 
+
+## Non-Goals
+
+* Chapter Two (getting the queries into our systems for our own
+  analytics) is explicitly out of scope of _this_ design, but is
+  necessary to unlock its full value, and so should be done as a
+  follow-up ASAP.
+* 100% durability of queries is permanently out of scope. We have to
+  accept the possibility of losing data under adverse conditions. The
+  reason for this design choice is that otherwise we would have to
+  persist the text of every statement (in either Cockroach or S3),
+  which we consider unacceptable for `SELECT` queries.
+
+## Overview
+
+The coordinator maintains a buffer of pending events related to
+statement execution. Whenever a prepared statement is created, a
+corresponding event is added to the buffer. Whenever a statement is
+executed, a system var called `STATEMENT_LOG_SAMPLING_RATE` is
+queried, and a random number between 0 and 1 is generated. If the
+random number is greater than the sampling rate, a description of the
+statement invocation is written into the buffer.
+
+Every so often (controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`), the
+buffer is converted to updates that are written out to
+`mz_statement_execution_history` and `mz_prepared_statement_history`,
+with a column in the former referencing the latter. These system
+tables are persistent, but entries older than `STATEMENT_LOG_MAX_AGE`
+are periodically deleted.
+
+## Detailed description
+
+The coordinator maintains a buffer of pending events, which can be of
+the following variants:
+
+``` rust
+enum StatementLogEvent {
+    StatementPrepared {
+        id: Uuid,
+        sql: String,
+        name: Option<String>,
+    },
+    StatementExecutionBegan {
+        id: Uuid,
+        prepared_statement_id: Uuid,        
+        params: Vec<Datum>,
+        when: DateTime,
+        // Additional metadata might eventually
+        // go here, e.g. `plan: Option<OptimizedMirRelationExpr>`
+    },
+    StatementExecutionFinished {
+        id: Uuid,
+        when: DateTime,
+        n_results: Option<usize>,
+    },
+}
+```
+
+The `StatementPrepared` and `StatementExecutionBegan` events are
+generated and stored in the buffer in `fn handle_execute`, after
+checking a random value against `STATEMENT_LOG_SAMPLING_RATE` . This
+requires keeping track of the prepared statement bound to each name,
+and whether it has been logged yet (as we don't want to log each
+statement several times). The `StatementExecutionFinished` events are
+generated in several different places, depending on the type of
+statement. For example, the event for `SELECT` queries is generated in
+response to the `ControllerResponse::PeekResponse` message.
+
```

Comment:
How will this work for `SUBSCRIBE` queries? They're a bit funny in that they don't _end_ in the coordinator, but rather the subscribe response channel gets handed back to the pgwire server.